frontend/utility/ThumbnailManager.cpp

```diff
@@ -0,0 +1,246 @@
+		auto &cachedItem = it->second;
+		pixmap = cachedItem.pixmap.value_or(QPixmap());
+		cachedItem.pixmap.reset();
+		cachedItem.weakActiveItem = weakActiveItem;
+	}
+}
+
+ThumbnailItem::~ThumbnailItem()
+{
+	auto thumbnailManager = OBSBasic::Get()->thumbnails();
+	if (!thumbnailManager) {
+		return;
+	}
+
+	auto &cachedItem = thumbnailManager->cachedThumbnails[uuid];
+	cachedItem.pixmap = pixmap;
+	cachedItem.weakActiveItem.clear();
+}
+
+void ThumbnailItem::imageUpdated(QImage image)
+{
+	QPixmap newPixmap;
+	if (!image.isNull()) {
+		newPixmap = QPixmap::fromImage(image);
+	}
+
+	pixmap = newPixmap;
+	emit updateThumbnail(pixmap);
+}
+
+void Thumbnail::thumbnailUpdated(QPixmap pixmap)
+{
+	emit updateThumbnail(pixmap);
+}
+
+ThumbnailManager::ThumbnailManager(QObject *parent) : QObject(parent)
+{
+	connect(&updateTimer, &QTimer::timeout, this, &ThumbnailManager::updateTick);
+}
+
+ThumbnailManager::~ThumbnailManager() {}
+
+QSharedPointer<Thumbnail> ThumbnailManager::getThumbnail(OBSSource source)
+{
+	std::string uuid = obs_source_get_uuid(source);
+
+	for (auto it = thumbnails.begin(); it != thumbnails.end(); ++it) {
+		auto item = it->toStrongRef();
+		if (item && item->uuid == uuid) {
+			return QSharedPointer<Thumbnail>::create(item);
+		}
+	}
+
+	QSharedPointer<Thumbnail> thumbnail;
+	if ((obs_source_get_output_flags(source) & OBS_SOURCE_VIDEO) != 0) {
+		auto item = QSharedPointer<ThumbnailItem>::create(uuid, source);
+		item->init(item.toWeakRef());
+
+		thumbnail = QSharedPointer<Thumbnail>::create(item);
+		connect(item.get(), &ThumbnailItem::updateThumbnail, thumbnail.get(), &Thumbnail::thumbnailUpdated);
+
+		newThumbnails.push_back(item.toWeakRef());
+	}
+
+	updateIntervalChanged(thumbnails.size());
+	return thumbnail;
+}
+
+bool ThumbnailManager::updatePixmap(QSharedPointer<ThumbnailItem> &sharedPointerItem)
+{
+	ThumbnailItem *item = sharedPointerItem.get();
+
+	OBSSource source = OBSGetStrongRef(item->weakSource);
+	if (!source) {
+		return true;
+	}
+
+	QPixmap pixmap;
+	item->pixmap = pixmap;
+
+	if (source) {
+		uint32_t sourceWidth = obs_source_get_width(source);
+		uint32_t sourceHeight = obs_source_get_height(source);
+
+		if (sourceWidth == 0 || sourceHeight == 0) {
+			return true;
+		}
+
+		auto obj = new ScreenshotObj(source);
+		obj->setSaveToFile(false);
+		obj->setSize(Thumbnail::cx, Thumbnail::cy);
+
+		connect(obj, &ScreenshotObj::imageReady, item, &ThumbnailItem::imageUpdated);
+	}
+
+	return true;
+}
+
+void ThumbnailManager::updateIntervalChanged(size_t newCount)
+{
+	int intervalMS = MIN_THUMBNAIL_UPDATE_INTERVAL_MS;
+	if (newThumbnails.size() == 0 && newCount > 0) {
+		int count = (int)newCount;
+		intervalMS = MIN_SOURCE_UPDATE_INTERVAL_MS / count;
+		if (intervalMS < MIN_THUMBNAIL_UPDATE_INTERVAL_MS) {
+			intervalMS = MIN_THUMBNAIL_UPDATE_INTERVAL_MS;
+		}
+	}
+
+	updateTimer.start(intervalMS);
+}
+
+void ThumbnailManager::updateTick()
+{
+	QSharedPointer<ThumbnailItem> item;
```

Comment:
Do not use QSharedPointer. Use std::shared_ptr instead.
See:
* https://github.com/obsproject/obs-studio/pull/9663